### PR TITLE
【AutoParallel】Add master grad in AMP-O2 of AutoParallel

### PR DESCRIPTION
**Pull request by heavyrain-lzy**
### PR types

New features
### PR changes

Others
### Description

Pcard-76459
Use the high precision gradient to replace the low precision gradient in optimizer to avoid inf/nan values of low precision.
The high precision gradient 'master grad' is used by `update_loss_scaling`, `GradClip` and `optimizer`.
You can enable the `master_grad` feature by set `use_master_grad=True` , default `False`.

The associated experiment taken on the single hose with 8 V-100 GPUs in the strategy `mp2pp4`:
image

### CODE REVIEW

File: python/paddle/distributed/passes/auto_parallel_fp16.py [hunk 1]
```diff
@@ -79,7 +79,7 @@ def set_auto_cast_attr(cast_op, block):
     ), f"in_var {in_name} or out_var {out_name} is None of cast op"
     if is_forward_op(cast_op):
         cast_op._set_attr('in_dtype', in_var.dtype)
-        cast_op._set_attr('out_dtype', out_var.dtype)
+        out_var.desc.set_dtype(paddle.dtype(cast_op.attr('out_dtype')))
```
Comment (zhiqiu) [inline python/paddle/distributed/passes/auto_parallel_fp16.py:82]: why change this?

File: python/paddle/distributed/passes/auto_parallel_master_grad.py [hunk 2]
```diff
@@ -0,0 +1,234 @@
...
+        if output_name in var_names:
+            grad_names.append(output_name)
+    return grad_names
+
+
+@register_pass("auto_parallel_master_grad_pass")
+class MasterGradPass(PassBase):
+    """
+    Use the high precision gradient to replace the low precision gradient in optimizer to avoid inf/nan values of low precision.
+    The high precision gradient 'master grad' is used by `update_loss_scaling`, `GradClip` and `optimizer`.
```
Comment (zhiqiu) [inline python/paddle/distributed/passes/auto_parallel_master_grad.py]: and communication?

File: python/paddle/distributed/passes/auto_parallel_master_grad.py [hunk 3]
```diff
@@ -0,0 +1,234 @@
...
+
+        # insert cast op
+        for grad_name, idx in reversed(grad_first_ids.items()):
+            grad_var = cur_block.var(grad_name)
+            if (
+                grad_var.dtype == core.VarDesc.VarType.FP16
+                or grad_var.dtype == core.VarDesc.VarType.BF16
+            ):
+                is_fp16 = grad_var.dtype == core.VarDesc.VarType.FP16
+                product_op = cur_block.ops[idx]
```
Comment (zhiqiu) [inline python/paddle/distributed/passes/auto_parallel_master_grad.py]: product -> producer

File: python/paddle/distributed/passes/auto_parallel_master_grad.py [hunk 4]
```diff
@@ -0,0 +1,234 @@
...
+    def _add_cast_op(self, cur_block, grad_names: List[str], dist_context):
+        grad_first_ids = OrderedDict()
+        for idx, op in enumerate(cur_block.ops):
+            if is_optimize_op(op):
+                break
+            elif is_backward_op(op):
+                var_names = get_output_in_varlist(op, grad_names)
+                for var_name in var_names:
+                    if var_name not in grad_first_ids:
+                        grad_first_ids[var_name] = idx
```
Comment (zhiqiu) [inline python/paddle/distributed/passes/auto_parallel_master_grad.py:86]: add comments on `else` branch

File: python/paddle/distributed/passes/auto_parallel_master_grad.py [hunk 5]
```diff
@@ -0,0 +1,234 @@
...
+                or grad_var.dtype == core.VarDesc.VarType.BF16
+            ):
+                is_fp16 = grad_var.dtype == core.VarDesc.VarType.FP16
+                product_op = cur_block.ops[idx]
+                product_op_dist_attr = (
+                    dist_context.get_op_dist_attr_for_program(product_op)
+                )
+                assert (
+                    product_op_dist_attr is not None
+                ), "The op should be distributed"
```
Comment (zhiqiu) [inline python/paddle/distributed/passes/auto_parallel_master_grad.py]: The op xx should be distributed

File: python/paddle/distributed/passes/auto_parallel_master_grad.py [hunk 6]
```diff
@@ -0,0 +1,234 @@
...
+                ref_output_dist_attr = (
+                    product_op_dist_attr.get_output_dist_attr(grad_name)
+                )
+                assert (
+                    ref_output_dist_attr is not None
+                ), "The output should be distributed"
```
Comment (zhiqiu) [inline python/paddle/distributed/passes/auto_parallel_master_grad.py]: same above

File: python/paddle/distributed/passes/auto_parallel_master_grad.py [hunk 7]
```diff
@@ -0,0 +1,234 @@
...
+                ref_mesh = ref_output_dist_attr.process_mesh
+                ref_dims_mapping = ref_output_dist_attr.dims_mapping
+                ref_chunk_id = product_op_dist_attr.chunk_id
+                grad_low_precision_name = (
```
Comment (zhiqiu) [inline python/paddle/distributed/passes/auto_parallel_master_grad.py]: half_precision

File: python/paddle/distributed/passes/auto_parallel_master_grad.py [hunk 8]
```diff
@@ -0,0 +1,234 @@
...
+        dist_context,
+    ):
+        grad_names = [g.name for _, g in params_grads]
+        # 1. delete the origin optimizer op
+        # 1.1 delete the var and op associated with the optimizer op in main_program
+        main_ops = main_program.global_block().ops
+        main_ops_len = len(main_ops)
+        first_optimize_idx = main_ops_len - 1
+        for idx, op in enumerate(main_ops):
+            if is_optimize_op(op) and main_ops[idx].type == "squared_l2_norm":
```
Comment (zhiqiu) [inline python/paddle/distributed/passes/auto_parallel_master_grad.py]: why? it seems too hard.

File: python/paddle/distributed/passes/auto_parallel_master_grad.py [hunk 9]
```diff
@@ -0,0 +1,234 @@
...
+                first_optimize_idx = idx
+                break
+        deleted_temp_var_names = []
+        deleted_persis_var_names = []
```
Comment (zhiqiu) [inline python/paddle/distributed/passes/auto_parallel_master_grad.py]: persis -> persist

File: python/paddle/distributed/passes/auto_parallel_master_grad.py [hunk 10]
```diff
@@ -0,0 +1,234 @@
...
+        reserved_var_names = []
+        for idx in range(main_ops_len - 1, first_optimize_idx - 1, -1):
+            op = main_ops[idx]
+            inout_arg_names = op.input_arg_names + op.output_arg_names
+            if op.type in _supported_optimizer_type:
+                param_names = op.input("Param")
+                learn_rate_names = op.input("LearningRate")
```
Comment (zhiqiu) [inline python/paddle/distributed/passes/auto_parallel_master_grad.py]: learning_rate

File: python/paddle/distributed/passes/auto_parallel_fp16.py (hunk 1 above)
Comment (heavyrain-lzy) [reply python/paddle/distributed/passes/auto_parallel_fp16.py:82]: This is a bug. We should set output by `out_dtype` of `cast` in forward pass.

File: python/paddle/distributed/passes/auto_parallel_master_grad.py (hunk 2 above)
Comment (heavyrain-lzy) [reply python/paddle/distributed/passes/auto_parallel_master_grad.py]: Done~

File: python/paddle/distributed/passes/auto_parallel_master_grad.py (hunk 4 above)
Comment (heavyrain-lzy) [reply python/paddle/distributed/passes/auto_parallel_master_grad.py:86]: Done~

File: python/paddle/distributed/passes/auto_parallel_master_grad.py (hunk 8 above)
Comment (heavyrain-lzy) [reply python/paddle/distributed/passes/auto_parallel_master_grad.py]: Done. I have changed the way to find the first optimizer operator excluding the operators for check_nan_inf.

**Review by zhiqiu — APPROVED** [review]
LGTM
